src/gateway_uri.rs

```diff
@@ -1,46 +1,147 @@
+    authority: Authority,
+}
 
 impl GatewayUri {
-    pub fn new(mut gateway_origin: Uri) -> Result<Self, BoxError> {
-        let (scheme, default_port) = match gateway_origin.scheme_str() {
-            Some("http") => ("http", 80),
-            Some("https") | None => ("https", 443),
-            _ => return Err("Unsupported URI scheme".into()),
+    pub fn new(scheme: Scheme, authority: Authority) -> Result<Self, BoxError> {
+        let default_port = if scheme == Scheme::HTTP {
+            80
+        } else if scheme == Scheme::HTTPS {
+            443
+        } else {
+            return Err("Unsupported URI scheme".into());
         };
 
-        if gateway_origin.authority().map(|a| a.port().is_none()).unwrap_or(true) {
-            let authority = if let Some(auth) = gateway_origin.authority() {
-                format!("{}:{}", auth.host(), default_port)
-            } else {
-                return Err("URI must have an authority".into());
-            };
+        // If no explicit port is provided, make the default one explicit
+        let mut authority = authority;
+        if authority.port().is_none() {
+            authority = Authority::from_str(&format!("{}:{}", authority.host(), default_port))
+                .expect("setting default port must succeed");
+        }
 
-            let path_and_query = gateway_origin
-                .path_and_query()
-                .map(|pq| pq.to_string())
-                .unwrap_or_else(|| "/".to_string());
+        Ok(Self { scheme, authority })
+    }
 
-            let builder =
-                Uri::builder().scheme(scheme).authority(authority).path_and_query(path_and_query);
+    // TODO remove
+    pub fn authority(&self) -> &Authority { &self.authority }
 
-            gateway_origin = builder.build()?;
-        }
+    pub fn from_static(string: &'static str) -> Self {
+        Uri::from_static(string).try_into().unwrap()
+    }
 
-        Ok(Self(gateway_origin))
+    fn to_uri_builder(&self) -> http::uri::Builder {
+        Uri::builder().scheme(self.scheme.clone()).authority(self.authority.clone())
     }
-}
 
-impl std::ops::Deref for GatewayUri {
-    type Target = Uri;
+    pub fn to_uri(&self) -> Uri {
+        self.to_uri_builder()
+            .path_and_query("/")
+            .build()
+            .expect("Building Uri from scheme and authority must succeed")
+    }
 
-    fn deref(&self) -> &Self::Target { &self.0 }
+    pub async fn to_socket_addr(&self) -> Option<std::net::SocketAddr> {
+        self.to_socket_addrs().await.ok()?.next()
+    }
+
+    pub async fn to_socket_addrs(
+        &self,
+    ) -> std::io::Result<impl Iterator<Item = std::net::SocketAddr>> {
+        tokio::net::lookup_host(self.authority.to_string()).await
+    }
 }
 
 impl From<GatewayUri> for Uri {
-    fn from(val: GatewayUri) -> Self { val.0 }
+    fn from(val: GatewayUri) -> Uri { val.to_uri() }
+}
+
+impl TryFrom<Uri> for GatewayUri {
+    type Error = BoxError;
```

Comment:
Is it not smelly to Box the error instead of creating a specialized type? As a way to make progress, I'm OK with it, but am wondering if you intend it to be permanent.